### DISCUSSION TITLE
SOLVED: trimesh integration failure on Windows

### DISCUSSION THREAD

**Discussion by villares** [py5coding/py5generator#556] [Q&A] [answered]
Hi people, I have this example that works fine for me on Linux, but it seems to have failed on a Windows machine at a school lab. It doesn't show any error but It doesn't show any image. :sob: 

![out](https://github.com/user-attachments/assets/524bee2b-0c80-4185-98d8-e31f204ee3d7)

```python
import shapely
import py5
import trimesh

def setup():
    global mesh, shp
    py5.size(600, 600, py5.P3D)
    p = shapely.Polygon([(100, 100), (0, 100), (100, 0)])
    # print(p)
    # print(p.centroid)
    # print(p.svg())    
    p_and_hole = p.buffer(50) - p
    mesh = trimesh.creation.extrude_polygon(p_and_hole, 100)
    shp = py5.convert_shape(mesh)

def draw():
    py5.background(100)
    py5.translate(py5.width  / 2, py5.height / 2)
    py5.rotate_y(py5.mouse_x / 10)
    py5.scale(1, -1)
    #shp.disable_style()
    #shp.set_fill(255)
    #py5.fill(255, 0, 200)
    py5.shape_mode(py5.CENTER)
    py5.shape(shp, 0, 0)
    
py5.run_sketch()

```

Can anyone check for me on Windows?

It was an older version o py5 from [the Thonny portable with  py5](https://github.com/villares/thonny-portable-with-py5), we might need to check on the latest version...

**Comment by hx2A** [comment]
I would double check the trimesh versions and the versions of every library trimesh depends on. Differences there might be causing this problem.

**Reply by vsquared** [reply]
Got this error on Windows 11:
ModuleNotFoundError: No module named 'shapely'
Installed both shapely and trimesh and it still won't run, but Windows is not my usual os.  Long error message:
```
>>> %Run trimeshDemo.py
try running `pip install mapbox-earcut manifold3d`or `triangle`, `mapbox_earcut`, then explicitly pass:
`triangulate_polygon(*args, engine="triangle")`
to use the non-FSF-approved-license triangle engine
py5 encountered an error in your code:

File "C:\Users\s\Python\trimeshDemo.py", line 13, in _py5_faux_setup
    5    def setup():
 (...)
    9        # print(p)
    10       # print(p.centroid)
    11       # print(p.svg())    
    12       p_and_hole = p.buffer(50) - p
--> 13       mesh = trimesh.creation.extrude_polygon(p_and_hole, 100)
    14       shp = py5.convert_shape(mesh)
    ..................................................
     trimesh.creation.extrude_polygon = <function 'extrude_polygon' creation.py:210>
     p_and_hole = <POLYGON ((150 0, 149.759 -4.901, 149.039 -9.755, 147.847 -1
                   4.514, 146.194 -...>
    ..................................................

File "C:\Users\s\Python\venv\lib\site-packages\trimesh\creation.py", line 235, in extrude_polygon
    210  def extrude_polygon(
    211      polygon: "Polygon", height: Number, transform: Optional[ArrayLike] = None, **kwargs
    212  ) -> Trimesh:
 (...)
    231      mesh : trimesh.Trimesh
    232        Resulting extrusion as watertight body
    233      """
    234      # create a triangulation from the polygon
--> 235      vertices, faces = triangulate_polygon(polygon, **kwargs)
    236      # extrude that triangulation along Z
    ..................................................
     polygon = <POLYGON ((150 0, 149.759 -4.901, 149.039 -9.755, 147.847 -1
                4.514, 146.194 -...>
     kwargs = {}
    ..................................................

File "C:\Users\s\Python\venv\lib\site-packages\trimesh\creation.py", line 666, in triangulate_polygon
    563  def triangulate_polygon(
    564      polygon,
    565      triangle_args: Optional[str] = None,
    566      engine: Optional[str] = None,
    567      force_vertices: bool = False,
    568      **kwargs,
    569  ) -> Tuple[NDArray[np.float64], NDArray[np.int64]]:
 (...)
    662              + "or `triangle`, `mapbox_earcut`, then explicitly pass:\n"
    663              + '`triangulate_polygon(*args, engine="triangle")`\n'
    664              + "to use the non-FSF-approved-license triangle engine"
    665          )
--> 666          raise ValueError("No available triangulation engine!")
    667  

ValueError: No available triangulation engine!
>>> 
```

**Reply by villares** [reply]
Cheers @vsquared, thank you for trying this for us! 
The error you got is still not the one we are investigating, it is just pointing that a library for the triangulation capabilities of trimesh is missing. If you also install `mapbox_earcut` (the one I usually install) it should go away. Another idea would be to install most of trimesh optional dependencies with `pip install trimesh[easy]`. Thanks again for yout time!

**Reply by vsquared** [reply]
pip to the rescue.  Thanks.  Sorry image is so large; I don't know how to shrink it.
![tm](https://github.com/user-attachments/assets/f4180e65-0666-4fc4-91ad-add15b05255f)

**Reply by villares** [reply]
Thank you @vsquared, it looks like it works perfectly for you. Good to know!

**Reply by villares** [reply]
> I would double check the trimesh versions and the versions of every library trimesh depends on. Differences there might be causing this problem.

Update: I was also able to run it on windows after updating shapely... let me see if I can manage to make a new portable and it might solve this :)

**Comment by villares** [comment, marked as answer]
Today I uploaded a new version of my Windows Thonny "portable" with updated versions of py5, shapely and trimesh pre-installed:

- https://github.com/villares/thonny-portable-with-py5/

I'll close this as solved for now, as it seems to work fine with the updated libraries, but I would appreciate any accounts of any other issues with it.

**Reply by vsquared** [reply]
The demo posted above does run without error in the portable Windows version, but is slow as noted in the Processing forum.

**Comment by villares** [comment]
I might want to  update my portable to update **jpype** to 1.5.1 :) 
https://github.com/jpype-project/jpype/issues/1111#issuecomment-2484390721

**Reply by villares** [reply]
Bad news, it looks like 1.5.1 breaks everything... 
![image](https://github.com/user-attachments/assets/4b9473ba-26f6-4536-a8f0-e6bb568ff363)
